internal/rbd/rbd_util.go

```diff
@@ -1165,13 +1225,42 @@ func (rv *rbdVolume) resize(newSize int64) error {
 	}
 	defer image.Close()
 
+	thick, err := rv.isThickProvisioned()
+	if err != nil {
+		return err
+	}
+
+	// offset is used to track from where on the expansion is done, so that
+	// the extents can be allocated in case the image is thick-provisioned
+	var offset uint64
+	if thick {
+		st, statErr := image.Stat()
+		if statErr != nil {
+			return statErr
+		}
+
+		offset = st.Size
+	}
+
 	err = image.Resize(uint64(util.RoundOffVolSize(newSize) * helpers.MiB))
 	if err != nil {
 		return err
 	}
 
+	if thick {
+		err = rv.allocate(offset)
```

Comment:
for expansion of thick image, do we need to first resize the image, allocate and again resize it?

Reply:
the steps are:

1. resize rbd image (by the provisioner)
2. allocate new extension (by the provisioner)
3. resize the filesystem (on the node)

In case step 2 fails, resize it again to the previous size, so that there is an error returned on the resize, without having additional allocations.

Reply:
currently, it does resize to the new image, and it's allocated to the old size. and if the allocation fails it again doing resize to the new size.

I was thinking it should be as follows

* Resize to the new size
* Get the image stat and allocate it to the new size
* If allocation fails shrink to the old size.

please correct me if am wrong here.

Reply:
I think that is done here? The early `Stat()` call is to get the pre-resize size of the image, so that it can be used as offset for writing zeros for the extension.